### PR TITLE
Mamba / FalconMamba: Fix mamba left padding

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

As pointed out in https://github.com/huggingface/transformers/pull/32080#discussion_r1705555054 - it is important to zero-out hidden states that corresponds to the padd tokens before and after the causal convolution so that the padd token will not have an impact on the calculated hidden states. 

This can be empirically proven by generation quality before / after this fix (note by default FalconMamba uses left padding):

```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

model_id = "tiiuae/falcon-mamba-7b"
tok = AutoTokenizer.from_pretrained(model_id)
tok.pad_token_id = tok.eos_token_id

texts = [
    "Hello today",
    "Hello my name is Younes and today"
]

inputs = tok(texts, return_tensors="pt", padding=True, return_token_type_ids=False).to(0)
model = AutoModelForCausalLM.from_pretrained(model_id, device_map=0, torch_dtype=torch.bfloat16)

out = model.generate(**inputs, max_new_tokens=20)
print(tok.batch_decode(out, skip_special_tokens=True))
```

### Before the fix:

> Hello today I'm.\nI'm.\n Hello today.\n Hello today.\n Hello today

### After the fix:

> Hello today I'm going to show you how to make a 3D model of a house.\n

Propagated the changes in Mamba1 as well

cc @ArthurZucker @molbap

### CODE REVIEW

File: src/transformers/models/falcon_mamba/modeling_falcon_mamba.py [hunk 1]
```diff
@@ -745,11 +773,17 @@ def prepare_inputs_for_generation(
         else:
             model_inputs = {"input_ids": input_ids.contiguous()}
 
+        # In case cache is not used, manually add a new column in the attention mask
+        if not use_cache and attention_mask is not None and input_ids.shape != attention_mask.shape:
+            padd_length = input_ids.shape[-1] - attention_mask.shape[-1]
+            attention_mask = torch.cat([attention_mask, torch.ones_like(input_ids[:, :padd_length])], dim=-1)
```
Comment (molbap) [inline src/transformers/models/falcon_mamba/modeling_falcon_mamba.py:779]: small typo!
```suggestion
        # In case cache is not used, manually add a new column in the attention mask
        if not use_cache and attention_mask is not None and input_ids.shape != attention_mask.shape:
            pad_length = input_ids.shape[-1] - attention_mask.shape[-1]
            attention_mask = torch.cat([attention_mask, torch.ones_like(input_ids[:, :pad_length])], dim=-1)
```

File: src/transformers/models/mamba/modeling_mamba.py [hunk 2]
```diff
@@ -261,6 +271,9 @@ def slow_forward(self, input_states, cache_params: Optional[MambaCache]=None, ca
             )
             hidden_states = self.act(self.conv1d(hidden_states)[..., :seq_len])         # [batch, intermediate_size, seq_len]
 
+        if attention_mask is not None:
+            hidden_states = hidden_states * attention_mask.unsqueeze(1)
```
Comment (molbap) [inline src/transformers/models/mamba/modeling_mamba.py:275]: valid for all of these operations: no `dtype` issue encountered here when the model is loaded in different precisions?

File: tests/models/falcon_mamba/test_modeling_falcon_mamba.py [hunk 3]
```diff
@@ -491,3 +494,23 @@ def test_generation_torch_compile(self):
...
+    def test_batched_generation(self):
+        model_id = "tiiuae/falcon-mamba-7b"
+        tok = AutoTokenizer.from_pretrained(model_id)
+        tok.pad_token_id = tok.eos_token_id
+
+        texts = ["Hello today", "Hello my name is Younes and today"]
+
+        EXPECTED_OUTPUT = [
+            "Hello today I'm going to show you how to make a 3D model of a house.\n",
+            "Hello my name is Younes and today I will be talking about the topic of “The importance of the internet in our life”.\n",
```
Comment (molbap) [inline tests/models/falcon_mamba/test_modeling_falcon_mamba.py:507]: :laughing: 
thanks for adding a test!

File: src/transformers/models/falcon_mamba/modeling_falcon_mamba.py [hunk 4]
```diff
@@ -733,6 +757,10 @@ def prepare_inputs_for_generation(
                 )
             if cache_position[0] > 0:
                 input_ids = input_ids[:, -1].unsqueeze(-1)
+
+                if attention_mask is not None:
+                    attention_mask = attention_mask[:, -1].unsqueeze(-1)
+
```
Comment (molbap) [inline src/transformers/models/falcon_mamba/modeling_falcon_mamba.py:763]: does this work if you generate with `inputs_embeds` as starter instead of `input_ids`?

**Review by molbap — COMMENTED** [review]
Thanks @younesbelkada for adding the states tuning-out! :grin: left a couple comments, mostly curious of some situations that were edge cases for mamba 2

File: src/transformers/models/mamba/modeling_mamba.py (hunk 2 above)
Comment (younesbelkada) [reply src/transformers/models/mamba/modeling_mamba.py:275]: In principle no since attention mask is a `torch.LongTensor`, in the slow test the model is loaded in bf16 and there is no issue !

File: src/transformers/models/falcon_mamba/modeling_falcon_mamba.py (hunk 4 above)
Comment (younesbelkada) [reply src/transformers/models/falcon_mamba/modeling_falcon_mamba.py:763]: Yes I think so as we will always want to get the last value of the attention mask for the last token 🤔

Comment (younesbelkada) [reply src/transformers/models/falcon_mamba/modeling_falcon_mamba.py:763]: Just tried it with https://github.com/huggingface/transformers/pull/32677/commits/8978e00919cec6f72c874a9762d25c292ea08d80 and I can confirm it works !

Comment (molbap) [reply src/transformers/models/falcon_mamba/modeling_falcon_mamba.py:763]: great, thanks!

**Comment by vasqu** [comment]
Can we propagate this to Jamba as well :D thx for this fix ❤️

**Review by molbap — APPROVED** [review]
LGTM! pinging @ArthurZucker for merging :slightly_smiling_face:

File: src/transformers/models/falcon_mamba/modeling_falcon_mamba.py [hunk 5]
```diff
@@ -733,6 +757,10 @@ def prepare_inputs_for_generation(
                 )
             if cache_position[0] > 0:
                 input_ids = input_ids[:, -1].unsqueeze(-1)
+
+                if attention_mask is not None:
+                    attention_mask = attention_mask[:, -1].unsqueeze(-1)
```
Comment (ArthurZucker) [inline src/transformers/models/falcon_mamba/modeling_falcon_mamba.py:762]:
```suggestion
                    attention_mask = None
```
actually we should not even pass a mask as it is always gonna be `1` here no?

File: src/transformers/models/falcon_mamba/modeling_falcon_mamba.py [hunk 6]
```diff
@@ -745,11 +773,17 @@ def prepare_inputs_for_generation(
         else:
             model_inputs = {"input_ids": input_ids.contiguous()}
 
+        # In case cache is not used, manually add a new column in the attention mask
+        if not use_cache and attention_mask is not None and input_ids.shape != attention_mask.shape:
+            pad_length = input_ids.shape[-1] - attention_mask.shape[-1]
```
Comment (ArthurZucker) [inline src/transformers/models/falcon_mamba/modeling_falcon_mamba.py:778]: this is not the pad length but the past_length rather

File: src/transformers/models/falcon_mamba/modeling_falcon_mamba.py [hunk 7]
```diff
@@ -745,11 +773,17 @@ def prepare_inputs_for_generation(
...
+            attention_mask = torch.cat([attention_mask, torch.ones_like(input_ids[:, :pad_length])], dim=-1)
```
Comment (ArthurZucker) [inline src/transformers/models/falcon_mamba/modeling_falcon_mamba.py:779]: Not sure I understand why we are adding a `[1] x batch_size`? ( past_length is usually gonna be `1 - current_generation_token` , so imagine 20 input ids, then -19 to slice the input_ids? 
Unless the inpud_ids is 20, but then it always has the same shape as the mask

**Review by ArthurZucker — COMMENTED** [review]
Thanks for adding a test 🤗

File: src/transformers/models/falcon_mamba/modeling_falcon_mamba.py (hunk 5 above)
Comment (younesbelkada) [reply src/transformers/models/falcon_mamba/modeling_falcon_mamba.py:762]: Good point !

File: src/transformers/models/falcon_mamba/modeling_falcon_mamba.py (hunk 6 above)
Comment (younesbelkada) [reply src/transformers/models/falcon_mamba/modeling_falcon_mamba.py]: Fixed

File: src/transformers/models/falcon_mamba/modeling_falcon_mamba.py (hunk 7 above)
Comment (younesbelkada) [reply src/transformers/models/falcon_mamba/modeling_falcon_mamba.py]: This is for users that run generation with `use_cache=False` and makes sure to manually update the attention mask because this is done no where else except here

Comment (ArthurZucker) [reply src/transformers/models/falcon_mamba/modeling_falcon_mamba.py]: then this is more a problem with `generate` as it should pass the correct attention mask 😓

**Review by ArthurZucker — APPROVED** [review]
Will include this in the patch 🤗

File: src/transformers/models/mamba/modeling_mamba.py [hunk 8]
```diff
@@ -557,6 +574,7 @@ def set_input_embeddings(self, new_embeddings):
     def forward(
         self,
         input_ids: Optional[torch.LongTensor] = None,
+        attention_mask: Optional[torch.LongTensor] = None,
```
Comment (ArthurZucker) [inline src/transformers/models/mamba/modeling_mamba.py:577]: this is breaking (having it as the second place)

Comment (younesbelkada) [reply src/transformers/models/mamba/modeling_mamba.py]: yes fixed it

File: src/transformers/models/mamba/modeling_mamba.py [hunk 9]
```diff
@@ -701,11 +729,17 @@ def prepare_inputs_for_generation(
         else:
             model_inputs = {"input_ids": input_ids.contiguous()}
 
+        # In case cache is not used, manually update the attention mask
+        if not use_cache and attention_mask is not None and input_ids.shape != attention_mask.shape:
+            past_length = input_ids.shape[-1] - attention_mask.shape[-1]
+            attention_mask = torch.cat([attention_mask, torch.ones_like(input_ids[:, :past_length])], dim=-1)
+
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/transformers/models/mamba/modeling_mamba.py]: Yes fixed it !

File: src/transformers/models/falcon_mamba/modeling_falcon_mamba.py [hunk 10]
```diff
@@ -735,6 +734,13 @@ def _update_model_kwargs_for_generation(
             and model_kwargs["cache_position"] is not None
         ):
             model_kwargs["cache_position"] = model_kwargs["cache_position"][-1:] + num_new_tokens
+
+        if "attention_mask" in model_kwargs:
+            attention_mask = model_kwargs["attention_mask"]
+            model_kwargs["attention_mask"] = torch.cat(
+                [attention_mask, attention_mask.new_ones((attention_mask.shape[0], 1))], dim=-1
+            )
```
Comment (ArthurZucker) [inline src/transformers/models/falcon_mamba/modeling_falcon_mamba.py:742]: good catch !

**Review by ArthurZucker — APPROVED** [review]
